### ISSUE TITLE
use RandomState for random numbers

### ISSUE THREAD

**Issue by josef-pkt** [statsmodels/statsmodels#306]
We should follow sklearn's example and use RandomState for any random variable or random process generation.

makes it easier to initialize with seed and will work better with parallel processing.

**Comment by josef-pkt** [comment]
revisit for 0.9 

scikit-learn and scipy use a helper function to transform user input which can be None, as seed or a RandomState instance.
(a bit similar to how we get matplotlib ax and figure keyword handling)

also we need a FAQ-D for it, as source for pattern and reminder. 
My multinomial proportion PR still uses global random state.

**Comment by josef-pkt** [comment]
another usecase: continue a permutation or monte carlo test, confidence interval (?) or similar

in an example for #3576, I used a generator sampling procedure with yield to be able to have an unlimited number of random samples.

However, if we can combine Monte Carlo results like for p-values (number of replication above a value), then we could just continue with the same RandomState in a new function call. The overhead in duplicate computation will in most cases be small relative to the computation for all monte carlo cases.

This might not work with updating confidence intervals or quantile unless we keep all previous monte carlo results available (i.e. online updating of e.g quantile for 5% largest and smallest values)

**Comment by josef-pkt** [comment]
related: this might be fragile

the code snippet below is inside a permutation hyopthesis test function, the underlying random sample generator is not vectorized (cannot be vectorized except with loop)
- `simulate_table_conditional(n_row, n_col, seed)` is a simple function. If we use the seed provided by the user in the main function, then we would call the random sample generator with the same seed each time. This will not work correctly. We would need to initialize a RandomState in the outer function if a seed was provided, and call the sampling function with the same RandomState instance.
- `simulate_table_permutation_gen(n_row, n_col, seed)` is a generator that is set up once with whatever RandomState initialization was used, and then we just iterate with `next`. Using a class that holds on to the RandomState would be similar.

```
    if 0:
        # this is slower than simple permutation generator 
        # (around 30 to 40% in simple example)
< see below >
        res_mc = [((simulate_table_conditional(n_row, n_col, seed) - count_ind)**2
                   / count_ind).sum()
                  for _ in range(n_repl)]
    else:
        gen = simulate_table_permutation_gen(n_row, n_col, seed)
        res_mc = [((next(gen) - count_ind)**2
                   / count_ind).sum()
                  for _ in range(n_repl)]
```

<...>  
This needs   
`rng = check_random_state(seed)`  
and then replace seed by rng in `simulate_table_conditional(n_row, n_col, rng)`

i.e. Never use a seed inside a sampling loop.

In a more elaborate version where we have vectorized chunking we would run into the same problems.

**Comment by jbrockmendel** [comment]
@bashtage does the reset_randomstate fixture make this closable?

**Comment by josef-pkt** [comment]
No, this is not for the unit tests. 
This is where we use random numbers in the regular code, e.g. for bootstrap, simulated (p-)values and randomized algorithms.

**Comment by josef-pkt** [comment]
for example the back-linked commit above 0fd2fd5fddf6c00d0cf3c09e42ce31615a0140ef
